modules/core/05-port/types/ibc_legacy_module.go

```diff
@@ -334,12 +334,76 @@ func (im *LegacyIBCModule) OnChanUpgradeTry(ctx sdk.Context, portID, channelID s
+		//
+		// If an application is a VersionWrapper which means it modifies the version string
+		// and the version string is non-empty (don't use default), then the application must
+		// attempt to unmarshal the version using the UnwrapVersionUnsafe interface function.
+		// If it is unsuccessful, no callback will occur to this application as the version
+		// indicates it should be disabled.
+		if wrapper, ok := im.cbs[i].(VersionWrapper); ok && strings.TrimSpace(counterpartyVersion) != "" {
+			appVersion, underlyingAppVersion, err := wrapper.UnwrapVersionUnsafe(counterpartyVersion)
+			if err != nil {
+				// middleware disabled
+				continue
+			}
+			cbVersion, counterpartyVersion = appVersion, underlyingAppVersion
+		}
+
+		// in order to maintain backwards compatibility, every callback in the stack must implement the UpgradableModule interface.
+		upgradableModule, ok := im.cbs[i].(UpgradableModule)
+		if !ok {
+			return errorsmod.Wrap(ErrInvalidRoute, "upgrade route not found to module in application callstack")
+		}
+
+		err := upgradableModule.OnChanUpgradeAck(ctx, portID, channelID, cbVersion)
+		if err != nil {
+			return errorsmod.Wrapf(err, "channel open init callback failed for port ID: %s, channel ID: %s", portID, channelID)
+		}
+	}
 	return nil
 }
 
 // OnChanUpgradeOpen implements the IBCModule interface
-func (LegacyIBCModule) OnChanUpgradeOpen(ctx sdk.Context, portID, channelID string, proposedOrder channeltypes.Order, proposedConnectionHops []string, proposedVersion string) {
+func (im *LegacyIBCModule) OnChanUpgradeOpen(ctx sdk.Context, portID, channelID string, proposedOrder channeltypes.Order, proposedConnectionHops []string, proposedVersion string) {
+	for i := len(im.cbs) - 1; i >= 0; i-- {
+		cbVersion := proposedVersion
+
+		// To maintain backwards compatibility, we must handle two cases:
+		// - relayer provides empty version (use default versions)
+		// - relayer provides version which chooses to not enable a middleware
+		//
+		// If an application is a VersionWrapper which means it modifies the version string
+		// and the version string is non-empty (don't use default), then the application must
+		// attempt to unmarshal the version using the UnwrapVersionUnsafe interface function.
+		// If it is unsuccessful, no callback will occur to this application as the version
+		// indicates it should be disabled.
+		if wrapper, ok := im.cbs[i].(VersionWrapper); ok && strings.TrimSpace(proposedVersion) != "" {
+			appVersion, underlyingAppVersion, err := wrapper.UnwrapVersionUnsafe(proposedVersion)
+			if err == nil {
+				cbVersion, proposedVersion = appVersion, underlyingAppVersion
+			} else {
+				// if there was an error unmarshalling the version, this should still be passed to the application.
+				// for example, it is possible to disable fee middleware by upgrading to a non fee incentivized
+				// channel.
+				cbVersion = proposedVersion
+			}
```

Comment:
would really appreciate some feedback on this change for correctness. 

This was introduced because we want to perform some logic in `OnChanUpgradeOpen` in the middleware, we don't want to skip it entirely.

For example, if we are upgrading from a fee transfer channel, to just a transfer channel, we need to `DeleteFeeEnabled`.

Reply:
The code should be the same as `OnChanOpenAck`

if `UnwrapVersionUnsafe` returns an error, we should give the module an empty string

Reply:
makes sense, then we can update the fee fn to

```golang
// OnChanUpgradeOpen implements the IBCModule interface
func (im IBCMiddleware) OnChanUpgradeOpen(ctx sdk.Context, portID, channelID string, proposedOrder channeltypes.Order, proposedConnectionHops []string, proposedVersion string) {
	if strings.TrimSpace(proposedVersion) == ""{
		// set fee disabled and pass through to the next middleware or application in callstack.
		im.keeper.DeleteFeeEnabled(ctx, portID, channelID)
		return
	}

	// set fee enabled and pass through to the next middleware of application in callstack.
	im.keeper.SetFeeEnabled(ctx, portID, channelID)
}
```